pkg/retriever/combinators/splitretriever.go

```diff
@@ -2,58 +2,61 @@ package combinators
 
 import (
 	"context"
+	"errors"
 
 	"github.com/filecoin-project/lassie/pkg/retriever/coordinators"
 	"github.com/filecoin-project/lassie/pkg/types"
 )
 
-var _ types.CandidateRetriever = SplitRetriever{}
+var _ types.CandidateRetriever = SplitRetriever[int]{}
 
 // SplitRetriever retrieves by first splitting candidates with a splitter, then passing the split sets
 // to multiple retrievers using the given coordination style
-type SplitRetriever struct {
-	CandidateSplitter   types.CandidateSplitter
-	CandidateRetrievers []types.CandidateRetriever
-	CoordinationKind    types.CoordinationKind
+type SplitRetriever[T comparable] struct {
+	AsyncCandidateSplitter types.AsyncCandidateSplitter[T]
+	CandidateRetrievers    map[T]types.CandidateRetriever
+	CoordinationKind       types.CoordinationKind
 }
 
-func (m SplitRetriever) Retrieve(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) types.CandidateRetrieval {
-	candidateRetrievals := make([]types.CandidateRetrieval, 0, len(m.CandidateRetrievers))
-	for _, candidateRetriever := range m.CandidateRetrievers {
-		candidateRetrievals = append(candidateRetrievals, candidateRetriever.Retrieve(ctx, request, events))
+func (m SplitRetriever[T]) Retrieve(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) types.CandidateRetreival {
+	candidateRetrievals := make(map[T]types.CandidateRetreival, len(m.CandidateRetrievers))
+	for key, candidateRetriever := range m.CandidateRetrievers {
+		candidateRetrievals[key] = candidateRetriever.Retrieve(ctx, request, events)
 	}
-	return splitRetrieval{
-		retrievalSplitter:   m.CandidateSplitter.SplitRetrieval(ctx, request, events),
+	return splitRetrieval[T]{
+		retrievalSplitter:   m.AsyncCandidateSplitter.SplitRetrievalRequest(ctx, request, events),
 		candidateRetrievals: candidateRetrievals,
 		coodinationKind:     m.CoordinationKind,
 		ctx:                 ctx,
 	}
 }
 
-type splitRetrieval struct {
-	retrievalSplitter   types.RetrievalSplitter
-	candidateRetrievals []types.CandidateRetrieval
+type splitRetrieval[T comparable] struct {
+	retrievalSplitter   types.AsyncRetrievalSplitter[T]
+	candidateRetrievals map[T]types.CandidateRetreival
 	coodinationKind     types.CoordinationKind
```

Comment:
```suggestion
	coordinationKind     types.CoordinationKind
```

would you mind changing the 3 instances of this while you're in there please?

Reply:
yea